test/gtest/uct/ib/test_ib_pkey.cc

```diff
@@ -66,36 +78,156 @@ class test_uct_ib_pkey : public test_uct_ib_with_specific_port {
     }
 
+    m_pkey_value[1] = m_pkey_value[0];
+    m_pkey_index[1] = m_pkey_index[0];
+
     send_recv_short();
 }
 
 UCS_TEST_P(test_uct_ib_pkey, all_avail_pkeys) {
     /* test all pkeys that are configured for the device */
     for (uint16_t table_idx = 0; table_idx < m_port_attr.pkey_tbl_len; table_idx++) {
-        m_pkey_value = query_pkey(table_idx);
-        if (!can_use_pkey(m_pkey_value)) {
+        m_pkey_value[0] = m_pkey_value[1] = query_pkey(table_idx);
+        if (!can_use_pkey(m_pkey_value[0])) {
             continue;
         }
         modify_config("IB_PKEY", "0x" +
-                      ucs::to_hex_string(m_pkey_value &
+                      ucs::to_hex_string(m_pkey_value[0] &
                                          UCT_IB_PKEY_PARTITION_MASK));
-        m_pkey_index = table_idx;
+        m_pkey_index[0] = m_pkey_index[1] = table_idx;
         send_recv_short();
     }
 }
 
+UCS_TEST_P(test_uct_ib_pkey, test_pkey_pairs) {
+    /* test all pkeys (even with limited membership) that are configured
+     * for the device */
+    ib_pkey_pairs_t pairs = supported_pkey_pairs(false);
+
+    for (size_t i = 0; i < pairs.first.size(); i++) {
+        m_pkey_value[0] = pairs.first[i][0];
+        m_pkey_value[1] = pairs.first[i][1];
+        m_pkey_index[0] = pairs.second[i][0];
+        m_pkey_index[1] = pairs.second[i][1];
+
+        modify_config("IB_PKEY", "0x" +
+                      ucs::to_hex_string(m_pkey_value[0] &
+                                         UCT_IB_PKEY_PARTITION_MASK));
+        m_e1 = uct_test::create_entity(0);
+        m_entities.push_back(m_e1);
+
+        modify_config("IB_PKEY", "0x" +
+                      ucs::to_hex_string(m_pkey_value[1] &
+                                         UCT_IB_PKEY_PARTITION_MASK));
+        m_e2 = uct_test::create_entity(0);
+        m_entities.push_back(m_e2);
+
+        prepare_entities();
+
+        uct_ib_iface_t *iface1 = ucs_derived_of(m_e1->iface(), uct_ib_iface_t);
+        uct_ib_iface_t *iface2 = ucs_derived_of(m_e2->iface(), uct_ib_iface_t);
+        uct_ib_address_t *ib_addr1 =
+            (uct_ib_address_t*)ucs_alloca(uct_ib_iface_address_size(iface1));
+        uct_ib_address_t *ib_addr2 =
+            (uct_ib_address_t*)ucs_alloca(uct_ib_iface_address_size(iface2));
+
+        /* pack-unpack the first IB iface address */
+        uint16_t lid1;
+        union ibv_gid gid1;
+        uint8_t gid_index1;
+        enum ibv_mtu mtu1;
+        uint16_t pkey_value1;
+        uct_ib_iface_address_pack(iface1, ib_addr1);
+        uct_ib_address_unpack(ib_addr1, &lid1, &gid1, &gid_index1, &mtu1,
+                              &pkey_value1);
+        if (m_pkey_value[0] != UCT_IB_PKEY_VALUE_DEFAULT) {
+            EXPECT_EQ(pkey_value1, m_pkey_value[0]);
+        } else {
+            EXPECT_EQ(0, pkey_value1);
+            pkey_value1 = UCT_IB_PKEY_VALUE_DEFAULT;
+        }
+
+        /* pack-unpack the second IB iface address */
+        uint16_t lid2;
```

Comment:
plz define some member func for common code

Reply:
done